src/blazar_tempest_plugin/tests/api/leases/test_lease_devices.py

```diff
@@ -0,0 +1,112 @@
+from datetime import datetime
+from datetime import timedelta
+
+from tempest.lib import exceptions
+from tempest.lib import decorators
+
+from blazar_tempest_plugin.common import utils
+from blazar_tempest_plugin.common import waiters
+from blazar_tempest_plugin.tests.api.base import ContainerApiBase
+
+
+class TestLeaseContainers(ContainerApiBase):
+    """Test leases for containers on chi@edge."""
+
+    def setUp(self):
+        super(TestLeaseContainers, self).setUp()
+        self.lease = self._reserve_device()
+        self.hints = {"reservation": utils.get_device_reservation_from_lease(self.lease)}
+        self.container = self._create_reserved_container("reservation-container", self.hints)
+        _, container = self.container_client.get_container(self.container.uuid)
+        self.assertEqual("Running", container.status)
+
+    @decorators.attr(type="smoke")
+    def test_extend_lease_for_reserved_container(self):
+        """Test extending a lease that has a reserved container."""
+
+        old_end = datetime.strptime(self.lease["end_date"], "%Y-%m-%dT%H:%M:%S.%f")
+        new_end = (old_end + timedelta(hours=2)).strftime("%Y-%m-%d %H:%M")
+
+        updated_lease = self.leases_client.update_lease(self.lease["id"], end_date=new_end)
+
+        final_lease = waiters.wait_for_lease_status(
+            self.leases_client, self.lease["id"], "ACTIVE"
+        )
+
+        expected_end_date = datetime.strptime(new_end, "%Y-%m-%d %H:%M")
+        actual_end_date = datetime.fromisoformat(final_lease["end_date"])
+
+        actual_end_date = actual_end_date.replace(microsecond=0)
+
+        self.assertEqual(
+            expected_end_date,
+            actual_end_date,
+            f"Expected end date {expected_end_date}, got {actual_end_date}"
+        )
+
+    @decorators.attr(type="smoke")
+    def test_delete_lease_with_container(self):
+        """Test deleting a lease with an associated container also deletes the container."""
+
+        _, container = self.container_client.get_container(self.container.uuid)
+        self.assertEqual("Running", container.status)
+
+        resp = self.leases_client.delete_lease(self.lease["id"])
+        self.assertEquals(200, resp.response.status)
+
+        waiters.wait_for_lease_termination(self.leases_client, self.lease["id"])
+
+        try:
+            self.container_client.ensure_container_in_desired_state(
+                self.container.uuid, "Deleted"
+            )
+        except exceptions.NotFound:
+            pass
```

Comment:
just leaving a future TODO / nit here, we should probably handle this in a helper method somewhere, because try/except in the test classes is an antipattern.

No worries about it for now.

Reply:
Not sure that I agree that using try in a test is always an antipattern :) In this case, I'd call it more of a "pragmatic workaround due to the eventually consistent nature of OpenStack APIs" :)

My bigger pet peeve of this code of mine is ignoring the exception, so I added a bit more to assert it more throughly.

Also happy to move this into a utility at some point as well.